{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalisation\n",
    "\n",
    "## Introduction\n",
    "\n",
    "In the previous section, we looked at estimating transcript abundance with Kallisto. The abundances are reported as **transcripts per million (TPM)**, but what does TPM mean and how is it calculated?\n",
    "\n",
    "The objectives of this part of the tutorial are:\n",
    "\n",
    "  * _understand why RNA-Seq normalisation metrics are used_\n",
    "  * _understand the difference between RPKM, FPKM and TPM_\n",
    "  * _calculate RPKM and TPM for a gene of interest_\n",
    "  \n",
    "There are many useful websites, publications and blog posts which go into much more detail about RNA-Seq normalisation methods. Here are just a couple (in no particular order):\n",
    "\n",
    "  * [What the FPKM? A review of RNA-Seq expression units](https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/)\n",
    "  * [RPKM, FPKM and TPM, clearly explained](https://statquest.org/2015/07/09/rpkm-fpkm-and-tpm-clearly-explained/)\n",
    "  * [A survey of best practices for RNA-seq data analysis](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4728800/)\n",
    "  * [The RNA-seq abundance zoo](http://robpatro.com/blog/?p=235)\n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Why do we use normalisation units instead of raw counts?\n",
    "\n",
    "Raw reads counts are the number of reads originating from each transcript which can be affected by several factors:\n",
    "\n",
    "  * **sequencing depth (total number of reads)**   \n",
    "    The more we sequence a sample, the more reads we expect to be assigned.   \n",
    "    \n",
    "     \n",
    "  * **gene/transcript length**    \n",
    "    The longer the gene or transcript, the more reads we expect to be assigned to it.    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Figure 4. Effect of sequencing depth and gene length on raw read counts](images/count-bias.png \"Effect of sequencing depth and gene length on raw read counts\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Look at the top part of Figure 4. In which sample, X or Y, is the gene more highly expressed?*\n",
    "\n",
    "Neither, it's the same in both. What we didn't tell you was that the total number of reads generated for sample A was twice the number than for sample B. That meant almost twice the number of reads are assigned to the same gene in sample A than in sample B.\n",
    "\n",
    "*Look at the bottom part of Figure 4. Which gene, X or Y, has the greatest gene level expression?*\n",
    "\n",
    "Neither, they are both expressed at the same level. This time we didn't tell you that gene X is twice the length of gene Y. This meant that almost twice the number reads were assigned to gene X than gene Y.\n",
    "\n",
    "In the top part of _Figure 4_, the gene in sample X has twice the number of reads assigned to it than the same gene in sample Y. What isn't shown is that sample X had twice the number or total reads than sample Y so we would expect more reads to be assigned in sample X. Thus, the gene is expressed at roughly the same level in both samples. In the bottom part of _Figure 4_, gene X has twice the number of reads assigned to it than gene Y. However, gene X is twice the length of gene Y and so we expect more reads to be assigned to gene X. Again, the expression level is roughly the same."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reads per kilobase per million (RPKM)\n",
    "\n",
    "Reads per kilobase (of exon) per million (reads mapped) or **RPKM** is a **within sample** normalisation method which takes into account sequencing depth and length biases.\n",
    "\n",
    "To calculate RPKM, you first normalise by sequencing depth and then by gene/transcript length.  \n",
    "\n",
    "  1. **Get your _per million_ scaling factor**  \n",
    "  Count up the total number of reads which have been assigned (mapped) in the sample. Divide this number by 1,000,000 (1 million) to get your *per million* scaling factor (N).\n",
    "  \n",
    "  2. **Normalise for sequencing depth**  \n",
    "  Divide the number of reads which have been assigned to the gene or transcript (C) by the _per million_ scaling factor you calculated in step 1. This will give you your reads per million (RPM).\n",
    "  \n",
    "  3. **Get your _per kilobase_ scaling factor**  \n",
    "  Divide the total length of the exons in your transcript or gene in base pairs by 1,000 (1 thousand) to get your _per kilobase_ scaling factor (L).\n",
    "  \n",
    "  4. **Normalise for length**  \n",
    "  Divide your RPM value from step 2 by your _per kilobase_ scaling factor (length of the gene/transcript in kilobases) from step 3. This will give you your reads per kilobase per million or RPKM.\n",
    "  \n",
    "This can be simplified into the following equation:\n",
    "\n",
    "\\begin{equation*}\n",
    "RPKM = \\frac {C}{LN}\n",
    "\\end{equation*}\n",
    " \n",
    " Where:\n",
    " \n",
    " * **C** is number of reads mapped to the transcript or gene\n",
    " * **L** is the total exon length of the transcript or gene in kilobases\n",
    " * **N** is the total number of reads mapped in millions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fragments per kilobase per million (FPKM)\n",
    "\n",
    "Fragments per kilobase per million or **FPKM** is essentially the same as RPKM except that: \n",
    "\n",
    "* RPKM is designed for **single-end** RNA-Seq experiments\n",
    "\n",
    "* FPKM is designed for **paired-end** RNA-Seq experiments\n",
    "\n",
    "In a paired-end RNA-Seq experiment, two reads may be assigned to a single fragment (in any orientation).  Also, in some cases, only one of those reads will be assigned to a fragment (singleton). The only difference between RPKM and FPKM is that FPKM takes into consideration that two reads may be assigned to the same fragment."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transcripts per million (TPM)\n",
    "\n",
    "Calculating the **transcripts per million** or **TPM** is a similar process to RPKM and FPKM. The main difference is that you will first normalise for length bias and then for sequencing depth bias. In a nut shell, we are swapping the order of normalisations.\n",
    "\n",
    "  1. **Get your _per kilobase_ scaling factor**  \n",
    "  Divide the total length of the exons in your transcript in base pairs by 1,000 (1 thousand) to get your _per kilobase_ scaling factor.  \n",
    "\n",
    "  2. **Normalise for length**  \n",
    "  Divide the number of reads which have been assigned to the transcript by the per kilobase scaling factor you calculated in step 1. This will give you your reads per kilobase (RPK).  \n",
    "  \n",
    "  3. **Get the sum of all RPK values in your sample**  \n",
    "  Calculate the RPK value for all of the transcripts in your sample. Add all of these together to get your total RPK value.   \n",
    "  \n",
    "  4. **Get your _per million_ scaling factor**  \n",
    "  Divide your total RPK value from step 3 by 1,000,000 (1 million) to get your _per million_ scaling factor.\n",
    "  \n",
    "  5. **Normalise for sequencing depth**  \n",
    "  Divide your RPK value calculated in step 2 by the _per million_ scaling factor from step 4. You now have your transcripts per millions value or TPM."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating RPKM and TPM values\n",
    "\n",
    "To try and answer this, let's look at a worked example. Here, we have three genes (A-C) and three biological replicates (1-3).\n",
    "\n",
    "| Gene | Length | Replicate 1 | Replicate 2 | Replicate 3 |\n",
    "| :-: | :-: | :-: | :-: | :-: |\n",
    "| **A** | 2,000 bases | 10 | 12 | 30 |\n",
    "| **B** | 4,000 bases | 20 | 25 | 60 |\n",
    "| **C** | 1,000 bases | 5 | 8 | 15 |\n",
    "\n",
    "\n",
    "There are two things to notice in our dataset:\n",
    "\n",
    "  * Gene B has twice number reads mapped than gene A, possibly as itâ€™s twice the length\n",
    " \n",
    "  \n",
    "  * Replicate 3 has more reads mapped than any of the other replicates, regardless of which gene we look at\n",
    "\n",
    "\n",
    "### Calculating RPKM\n",
    "\n",
    "#### Step 1:  get your per million scaling factor\n",
    "\n",
    "In the table below is the total number of reads which mapped for each of the replicates. To get our _per million_ scaling factor, we divide each of these values by 1,000,000 (1 million).\n",
    "\n",
    "| Gene | Replicate 1 | Replicate 2 | Replicate 3 |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| Total reads mapped | 3,500,000 | 4,500,000 | 10,600,000 |\n",
    "| Per million reads | 3.5 | 4.5 | 10.6 |\n",
    "\n",
    "\n",
    "#### Step 2:  normalise for sequencing depth\n",
    "\n",
    "We now divide our read counts by the _per million_ scaling factor to get our reads per million (RPM).\n",
    "\n",
    "Before:  \n",
    "\n",
    "| Gene | Replicate 1 | Replicate 2 | Replicate 3 |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| **A** | 10 | 12 | 30 |\n",
    "| **B** | 2 | 25 | 60 |\n",
    "| **C** | 5 | 8 | 15 |\n",
    "\n",
    "\n",
    "After:  \n",
    "\n",
    "| Gene | Replicate 1 RPM | Replicate 2 RPM | Replicate 3 RPM |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| **A** | 2.857 | 2.667 | 2.830 |\n",
    "| **B** | 5.714 | 5.556 | 5.660 |\n",
    "| **C** | 1.429 | 1.778 | 1.415 |\n",
    "\n",
    "\n",
    "#### Step 3: get your _per kilobase_ scaling factor\n",
    "\n",
    "Here we have our gene length in base pairs. For our _per kilobase_ scaling factor we need to get our gene length in kilobases by dividing it by 1,000.\n",
    "\n",
    "| Gene | Length (base pairs)| Length (kilobases) |\n",
    "| :-: | :-: | :-: |\n",
    "| **A** | 2,000 | 2 |\n",
    "| **B** | 4,000 | 4 |\n",
    "| **C** | 1,000 | 1 |\n",
    "\n",
    "#### Step 4: normalise for length\n",
    "\n",
    "Finally, we divide our RPM values from step 2 by our _per kilobase_ scaling factor from step 3 to get our reads per kilobase per million (RPKM).\n",
    "\n",
    "Before:\n",
    "\n",
    "| Gene | Replicate 1 RPM | Replicate 2 RPM | Replicate 3 RPM |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| **A** | 2.857 | 2.667 | 2.830 |\n",
    "| **B** | 5.714 | 5.556 | 5.660 |\n",
    "| **C** | 1.429 | 1.778 | 1.415 |\n",
    "\n",
    "After:\n",
    "\n",
    "| Gene | Replicate 1 RPKM | Replicate 2 RPKM | Replicate 3 RPKM |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| **A** | 1.43 | 1.33 | 1.42 |\n",
    "| **B** | 1.43 | 1.39 | 1.42 |\n",
    "| **C** | 1.43 | 1.78 | 1.42 |\n",
    "\n",
    "\n",
    "Notice that even though replicate 3 had more reads assigned than the other samples and a greater sequencing depth, its RPKM is quite similar. And, that although gene B had twice the number of reads assigned than gene A, its RPKM is the same. This is because we have normalised by both length and sequencing depth."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculating TPM\n",
    "\n",
    "Now we're going to calculate the TPM values for the same example data. As a reminder, here are our three genes (A-C) and three biological replicates (1-3).\n",
    "\n",
    "| Gene | Length | Replicate 1 | Replicate 2 | Replicate 3 |\n",
    "| :-: | :-: | :-: | :-: | :-: |\n",
    "| **A** | 2,000 bases | 10 | 12 | 30 |\n",
    "| **B** | 4,000 bases | 20 | 25 | 60 |\n",
    "| **C** | 1,000 bases | 5 | 8 | 15 |\n",
    "\n",
    "#### Step 1: get your _per kilobase_ scaling factor\n",
    "\n",
    "Again, our gene lengths are in base pairs. For our _per kilobase_ scaling factor we need to get our gene length in kilobases by dividing it by 1,000.\n",
    "\n",
    "| Gene | Length (base pairs)| Length (kilobases) |\n",
    "| :-: | :-: | :-: |\n",
    "| **A** | 2,000 | 2 |\n",
    "| **B** | 4,000 | 4 |\n",
    "| **C** | 1,000 | 1 |\n",
    "\n",
    "#### Step 2: normalise for length\n",
    "\n",
    "Now we divide the number of reads which have been assigned to each gene by the per kilobase scaling factor we just calculated. This will give us our reads per kilobase (RPK).  \n",
    "\n",
    "Before:  \n",
    "\n",
    "| Gene | Replicate 1 | Replicate 2 | Replicate 3 |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| **A** | 10 | 12 | 30 |\n",
    "| **B** | 2 | 25 | 60 |\n",
    "| **C** | 5 | 8 | 15 |\n",
    "\n",
    "After:\n",
    "\n",
    "| Gene | Replicate 1 | Replicate 2 | Replicate 3 |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| **A** | 5 | 6 | 15 |\n",
    "| **B** | 5 | 6.25 | 15 |\n",
    "| **C** | 5 | 8 | 15 |\n",
    "\n",
    "\n",
    "#### Step 3: get the sum of all RPK values in your sample\n",
    "\n",
    "Next, we sum the RPK values for each of our replices. This will give use our total RPK value for each replicate. To make this example scalable, we assume there are other genes so the total RPK is made up.\n",
    "\n",
    "| Gene | Replicate 1 | Replicate 2 | Replicate 3 |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| **A** | 5 | 6 | 15 |\n",
    "| **B** | 5 | 6.25 | 15 |\n",
    "| **C** | 5 | 8 | 15 |\n",
    "| ... | ... | ... | ... |\n",
    "| **Total RPK** | **150,000** | **202,500** | **352,500** |\n",
    "\n",
    "#### Step 4: get your _per million_ scaling factor\n",
    "\n",
    "Here, instead of dividing our total mapped reads by 1,000,000 (1 million) to get our _per million_ scaling factor, we divide our total RPK values by 1,000,000 (1 million).\n",
    "\n",
    "| Gene | Replicate 1 | Replicate 2 | Replicate 3 |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| Total RPK | 150,000 | 202,500 | 352,500 |\n",
    "| Per million RPK | 0.1500 | 0.2025 | 0.3525 |\n",
    "\n",
    "#### Step 5: normalise for sequencing depth\n",
    "\n",
    "Finally, we divide our individual RPK values from step 2 by the _per million_ scaling factor in step 4 to give us our TPM values.\n",
    "\n",
    "Before:\n",
    "\n",
    "| Gene | Replicate 1 | Replicate 2 | Replicate 3 |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| **A** | 5 | 6 | 15 |\n",
    "| **B** | 5 | 6.25 | 15 |\n",
    "| **C** | 5 | 8 | 15 |\n",
    "\n",
    "After:\n",
    "\n",
    "| Gene | Replicate 1 | Replicate 2 | Replicate 3 |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| **A** | 33.33 | 29.63 | 31.21 |\n",
    "| **B** | 33.33 | 30.86 | 31.91 |\n",
    "| **C** | 33.33 | 39.51 | 36.88 |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Which normalisation unit should I use?\n",
    "\n",
    "Well, there's a lot of debate around this, so let's look at our total normalised values for each replicate.\n",
    "\n",
    "### RPKM\n",
    "\n",
    "| Gene | Replicate 1 RPKM | Replicate 2 RPKM | Replicate 3 RPKM |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| **A** | 1.43 | 1.33 | 1.42 |\n",
    "| **B** | 1.43 | 1.39 | 1.42 |\n",
    "| **C** | 1.43 | 1.78 | 1.42 |\n",
    "| **Total RPKM** | **4.29** | **4.50** | **4.25** |\n",
    "\n",
    "### TPM\n",
    "\n",
    "| Gene | Replicate 1 | Replicate 2 | Replicate 3 |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| **A** | 33.33 | 29.63 | 31.21 |\n",
    "| **B** | 33.33 | 30.86 | 31.91 |\n",
    "| **C** | 33.33 | 39.51 | 36.88 |\n",
    "| **Total TPM** | **100** | **100** | **100** |\n",
    "\n",
    "\n",
    "Notice that that total TPM value for each of the replicates is the same. This is not true for RPKM and FPKM where the total values differ. With TPM, having the same total value for each replicate makes it easier to compare the proportion of reads mapping to each gene across replicates (although you shouldn't really compare across experiments).  With RPKM and FPKM, the differing total values make it much harder to compare replicates."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions\n",
    "\n",
    "Below is the information for each of the five samples. You will need this information to answer the questions. We have put all of commands used to get this information in the [answers](aanswers.ipynb).\n",
    "\n",
    "| Sample | Total mapped reads | Transcript length | Assigned reads | Total RPK |\n",
    "| :-: | :-: | :-: | :-: | :-: |\n",
    "| MT1 | 2,353,750 | 3,697 | 2,541 | 293,431 |\n",
    "| MT2 | 2,292,271 | 3,709 | 3,392 | 675,190 |\n",
    "| SBP1 | 2,329,235 | 3,699 | 14,605 | 1,719,970 |\n",
    "| SBP2 | 2,187,718 | 3,696 | 17,302 | 1,429,540 |\n",
    "| SBP3 | 2,163,979 | 3,699 | 14,646 | 1,561,310 |\n",
    "\n",
    "_Note: values have been rounded up to integers to make calculations easier. Assigned reads are the est_count from Kallisto for PCHAS_1402500. Transcript lengths are the est_length from Kallisto for PCHAS_1402500._\n",
    "\n",
    "**Q1: Using the `abundance.tsv` files generated by Kallisto and the information above, calculate the RPKM for PCHAS_1402500 in each of our five samples.**\n",
    "\n",
    "| Sample | Per million scaling factor | RPM | Per kilobase scaling factor | RPKM |\n",
    "| :-: | :-: | :-: | :-: | :-: |\n",
    "| MT1 |  |  |  |  |\n",
    "| MT2 |  |  |  |  |\n",
    "| SBP1 |  |  |  |  |\n",
    "| SBP2 |  |  |  |  |\n",
    "| SBP3 |  |  |  |  |\n",
    "\n",
    "**Q2: Using the `abundance.tsv` files generated by Kallisto and the information above, calculate the TPM for PCHAS_1402500 in each of our five samples.**  \n",
    "_Hint: don't forget to get your per million scaling factor._\n",
    "\n",
    "| Sample | Per kilobase scaling factor | Reads per kilobase (RPK) | TPM |\n",
    "| :-: | :-: | :-: | :-: |\n",
    "| MT1 |  |  |  |\n",
    "| MT2 |  |  |  |\n",
    "| SBP1 |  |  |  |\n",
    "| SBP2 |  |  |  |\n",
    "| SBP3 |  |  |  |\n",
    "\n",
    "**Q3: Do these match the TPM values from Kallisto?**  \n",
    "_Hint: look at the `abundance.tsv` files for each of your samples._\n",
    "\n",
    "**Q4: Do you think PCHAS_1402500 is differentially expressed between the MT and SBP samples?**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What's next?\n",
    "\n",
    "You can head back to **[transcript quantification with Kallisto](transcript-quantification.ipynb) ** or continue on to  **[Identifying differentially expressed genes with Sleuth](sleuth-de.ipynb) **."
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
